### PR TITLE
Api quota

### PR DESCRIPTION
**Pull request by alex-aipolabs**
### 🏷️ Ticket

[notion](https://www.notion.so/fbce3fc996ad4b2aaacd85d0492a48d7?v=c135b6e7f76243819caf99a83e291380&p=2088378d6a4780a7a28cc558d08cfd04&pm=s)

### 📝 Description

[Describe your changes in detail (optional if the issue you linked already contains a
detail description of the change)]

### ✅ Checklist

- [ ] I have signed the [Contributor License Agreement]() (CLA) and read the [contributing guide](./../CONTRIBUTING.md) (required)
- [ ] I have linked this PR to an issue or a ticket (required)
- [ ] I have updated the documentation related to my change if needed
- [ ] I have updated the tests accordingly (required for a bug fix or a new feature)
- [ ] All checks on CI passed

## Summary by CodeRabbit

- **New Features**
  - Introduced monthly API quota tracking and enforcement per project and organization, with automatic reset at the start of each billing period.
  - Users can now view their monthly API quota usage and last reset date in project and subscription details.
  - Subscription details now display the start of the current billing period.

- **Bug Fixes**
  - Ensured quota resets correctly when billing periods change or new subscriptions are created.

- **Tests**
  - Added comprehensive tests for monthly API quota increments, resets, enforcement, and edge cases.

- **Chores**
  - Updated test fixtures and parameter names for improved clarity and consistency.

### CODE REVIEW

File: backend/aci/server/billing.py [hunk 1]
```diff
@@ -26,5 +34,81 @@ def get_subscription_by_org_id(db_session: Session, org_id: UUID) -> Subscriptio
         raise SubscriptionPlanNotFound(f"Plan {subscription.plan_id} not found")
     return SubscriptionFiltered(
         plan=plan,
+        current_period_start=subscription.current_period_start,
         status=subscription.status,
     )
+
+
+def increase_quota_usage(db_session: Session, project: Project) -> None:
```
Comment (thisisfixer) [inline backend/aci/server/billing.py:42]: i think it's better to have your function do just one thing.
the reset & increment & check success mixed together is a bit confusing

File: backend/aci/server/billing.py [hunk 2]
```diff
@@ -26,5 +34,71 @@ def get_subscription_by_org_id(db_session: Session, org_id: UUID) -> Subscriptio
...
+                "monthly_quota_limit": monthly_quota_limit,
+            },
+        )
+        raise MonthlyQuotaExceeded(
+            f"monthly quota exceeded for org={project.org_id}, "
+            f"usage={total_monthly_usage}, limit={monthly_quota_limit}"
+        )
+
+
+def increment_quota_or_reset(db_session: Session, project: Project) -> None:
```
Comment (thisisfixer) [inline backend/aci/server/billing.py:91]: `increment_quota_or_reset ` this function is not actually doing a `or`?
what it does is either `increment_quota but NOT reset`, `increment_quota AND reset`

maybe remove this function from this file, and have a dependency function in dependency file called validate_org_quota (similiar to the old validate_project_quota), and move the implementation of this function to that dependency function

File: backend/aci/server/billing.py [hunk 3]
```diff
@@ -26,5 +34,71 @@ def get_subscription_by_org_id(db_session: Session, org_id: UUID) -> Subscriptio
         raise SubscriptionPlanNotFound(f"Plan {subscription.plan_id} not found")
     return SubscriptionFiltered(
         plan=plan,
+        current_period_start=subscription.current_period_start,
         status=subscription.status,
     )
+
+
+def reset_quota_if_period_changed(
```
Comment (thisisfixer) [inline backend/aci/server/billing.py:42]: im still a bit confused how the reset works, and what do you mean by `billing period has changed`?
- does it mean when user changed their billing plan (or cancel etc)
- or does it mean when it's a new month? if this is the case it shouldn't be called `billing period` changed?

and for free plan, when does the reset happen?

**Review by thisisfixer — CHANGES_REQUESTED** [review]

File: backend/aci/server/billing.py [hunk 4]
```diff
@@ -17,6 +22,9 @@ def get_subscription_by_org_id(db_session: Session, org_id: UUID) -> Subscriptio
             raise SubscriptionPlanNotFound("Free plan not found")
         return SubscriptionFiltered(
             plan=plan,
+            current_period_start=datetime.now(UTC).replace(
```
Comment (thisisfixer) [inline backend/aci/server/billing.py:25]: can remove current_period_start if no longer needed

File: backend/aci/server/billing.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
-from datetime import UTC
 from uuid import UUID
 
+from pytz import UTC
```
Comment (thisisfixer) [inline backend/aci/server/billing.py:3]: what's the reason for using pytz UTC instead of datetime?

File: backend/aci/server/tests/conftest.py [hunk 6]
```diff
@@ -3,21 +3,24 @@
...
 from typing import cast
 from unittest.mock import patch
 
 import pytest
+from dateutil.relativedelta import relativedelta
 from fastapi.testclient import TestClient
 from propelauth_fastapi import User
 from propelauth_py.types.login_method import SocialLoginProvider, SocialSsoLoginMethod
 from propelauth_py.types.user import OrgMemberInfo
+from pytz import UTC
```
Comment (thisisfixer) [inline backend/aci/server/tests/conftest.py:16]: same here `pytz `

File: backend/aci/server/dependencies.py [hunk 7]
```diff
@@ -133,6 +134,15 @@ def validate_monthly_api_quota(
     2. Reset quota if new billing cycle has started
     3. Increment usage or raise error if exceeded
     """
+    # Only check quota for app search and function search/execute endpoints
+    path = request.url.path
+    is_valid_endpoint = path.startswith(f"{config.ROUTER_PREFIX_APPS}/search") or (
```
Comment (thisisfixer) [inline backend/aci/server/dependencies.py:139]: not trying to be picky here and it's ok to leave it as it is. But when coming up var names let's have a clear, unambiguous name.
in this case it's endpoints that have quota limitations, "valid" endpoints could mean something very different

File: backend/aci/server/tests/conftest.py [hunk 8]
```diff
@@ -773,4 +776,62 @@ def starter_plan(db_session: Session) -> Plan:
...
+        dummy_free_plan: The free plan fixture
+        dummy_starter_plan: The starter plan fixture
+        dummy_project_1: The project to create the subscription for
+        db_session: The database session
+
+    Returns:
+        A Subscription object for paid plans, or None for the free plan
+    """
+    # Get plan type from request parameter, default to STARTER
+    plan_type = request.param or PlanType.STARTER
```
Comment (thisisfixer) [inline backend/aci/server/tests/conftest.py:804]: although it's just a testcase fixture, i have to read the code implementation to know what plan im getting when i make a call to dummy_subscription (that information should be clear on the interface, which is the function signature).
you can leave this as it is, but imo many times clarity matters more than having less duplicate code.

**Review by thisisfixer — CHANGES_REQUESTED** [review]

File: backend/aci/server/billing.py [hunk 9]
```diff
@@ -1,12 +1,17 @@
+from datetime import UTC
```
[the remark this answers is not in the document]
Comment (alex-aipolabs) [reply backend/aci/server/billing.py:1]: since we have specified the version in the project i think its ok to use UTC from datetime

File: backend/aci/server/tests/conftest.py (hunk 8 above)
Comment (alex-aipolabs) [reply backend/aci/server/tests/conftest.py:803]: for the subscription this fixture only works if you use `@pytest.mark.parametrize("dummy_subscription", [PlanType.FREE, PlanType.STARTER], indirect=True)`  on the class or the test function. I thought this will be better so whenever we use the subscription we have to test for both free and starter plans
